### PR TITLE
mir.random: initial API

### PR DESCRIPTION
**Pull request by wilzbach**
Hey,

so I started to think (and play) with a new random API.
Please don't invest time looking over the code - it's just a draft and I want to use to ask better questions.
Update: only have a [lookat the unittest at the bottom](https://github.com/DlangScience/mir/pull/85/files#diff-a1906e86b965bd2806454874420b9c70R137)

Questions
1) Do we want methods to use a `struct` or just a `function` for the random generator methods? 
If you have a peek at [Boost::Random](https://github.com/boostorg/random), you will see that [`reset`](https://github.com/boostorg/random/blob/develop/include/boost/random/gamma_distribution.hpp#L155) is quite common. However [dstats](https://github.com/dsimcha/dstats/blob/master/source/dstats/random.d) avoids general states, but stores them if needed as a [singleton](https://github.com/dsimcha/dstats/blob/master/source/dstats/random.d#L485).

2) Do we want to save a reference to the random generator in such a struct?
The advantage is that when passing the reference, the struct is an usual infinite forward Range (see 3 too).

3) If the struct gets duplicates via `save` - should the random generator get duplicated too? A la

```
assert(randomMethod.front.approxEqual(0.786));
assert(randomMethod.save.dropOne.front.approxEqual(0.831));
assert(randomMethod.front.approxEqual(0.786));
```

4) The names are still horrible, but do you prefer to manual construction of a `struct` like `rarray(Bernoulli(0.5), 10)` or having a wrapper around it `rarray!Bernoulli(10, 0.5)` or both styles?

### CODE REVIEW

**Comment by 9il** [comment]
1. No range interfaces

Function should accept RNG and return value

**Comment by 9il** [comment]
... and state if required

File: source/mir/random/package.d [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+/*
+WIP - Currently only
+$(WEB https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform, Box Muller-transform)
+References:
+    Butcher, J. C. "Random sampling from the normal distribution."
+    The Computer Journal 3.4 (1961): 251-253.
+    http://comjnl.oxfordjournals.org/content/3/4/251.full.pdf
+*/
+auto normal(RGen)(double mu, double sigma, ref RGen gen = rndGen)
```
Comment (9il) [inline source/mir/random/package.d:147]: we need compression of different normal RNGs

Comment (9il) [inline source/mir/random/package.d:147]: I found 3 API errors in this function... we just need good common C style functions. API is the last step for this module

Comment (wilzbach) [inline source/mir/random/package.d:147]: do you mean compression or comparison?

> we just need good common C style functions. API is the last step for this module

ok!

Comment (9il) [inline source/mir/random/package.d:147]: > I found 3 API errors in this function... we just need good common C style functions. API is the last step for this module

comparison))) There are different normal RNG and we should implement the best or few best algorithms

**Comment by 9il** [comment]
1. structs
2. yes
3. no
4. see mir.random.algorithm in mir-random

Should be reopened for mir-random

**Pull request closed without merging**